### ISSUE TITLE
make! inside rspec let not being reset along with test fixture transactional state

### ISSUE THREAD

**Issue by knzai** [notahat/machinist#93]
I've been adding Machinists to a project with existing fixtures.  We have transactional_fixtures turned on and I'm using rspec's let to set some of the factories

``` ruby
describe MyModel
  let(:recent_object) { MyModel.make!(:end_date => 2.days.ago) }

  it "does this thing"
    something_that_references_recent_object
    recent_object.should do_this_thing
  end
end
```

These objects seem to be persisting outside of the scope of the usual transaction, even though they are getting initially referenced (and thus instantiated) inside of it blocks.  If I force their instantiation using let! they seem to be properly handled within the transaction.  This is annoying because I'm counting on the laziness of let so I can set them all up and use just the ones I want.

I can also set-up more nested context's with before blocks to handle this, but it seems like perhaps something is just falling through the cracks.

Do I need to dive into the guts of rspec a little to see what is going on, or is this just machinist not quite handling let properly?

**Comment by TylerRick** [comment]
I think I ran into this same problem recently, or one similar to it anyway. I ended up turning `transactional_fixtures` off and that seemed to help. I never figured out what was going on though.

I just found my notes for when I had the problem:

``` ruby
    before do
      #puts "OrgEntity.all=#{OrgEntity.all.inspect}"

      # When we had config.use_transactional_fixtures = true,
      # This resulted in the insert (make!) causing a deadlock (I'm guessing?) and timing out!!!:
      #OrgEntity.delete_all

      # This didn't work so well either:
      #OrgEntity.destroy_all

      # But this works. Why is this???
      # And where is that extra record coming from??
      DatabaseCleaner.clean

      #puts "OrgEntity.all=#{OrgEntity.all.inspect}"
    end

    it 'orders by name' do
      b = OrgEntity.make!(name: 'b')
      a = OrgEntity.make!(name: 'a')
      OrgEntity.all.should == [a, b]
    end
```

I guess I wasn't using `let`, but I was using make! and `transactional_fixtures`. Sorry if this is unrelated.

(machinist 2)

**Comment by knzai** [comment]
Yeah, you can fix it by turning off transactional fixtures and using DatabaseCleaner, but that's not a fun solution if you are migrating a large project with lots of fixtures to machinist.

**Comment by pat** [comment]
I'm pretty sure this is fixed in the latest commits - you're seeing the remnants of the attempted object caching, which has been removed.

**Comment by knzai** [comment]
I am pretty sure it was running it off of HEAD from after all the object cacheing was removed. The date on the ticket was when I was hitting the bug, and I think I was using git master as the bundler source.

I'll look at the codebase in question and try and verify this week.

**Comment by lawitschka** [comment]
This ticket seems to be fixed. I experienced these problems, too, and just fetched HEAD instead of the Beta 2 version and transactional fixtures are now working with `let` without the usage of DatabaseCleaner.

**Comment by TylerRick** [comment]
I ran into (what I believe is) this same issue again with cucumber.

When I switched from `DatabaseCleaner.strategy = :truncation` to `DatabaseCleaner.strategy = :transaction`, I was getting weird results when I did `Model.make` -- associated records created by that record were not getting saved or something like that. (And that was all working fine when I had `DatabaseCleaner.strategy = :truncation`, but I needed to switch to `:transaction` for other reasons....)

`DatabaseCleaner`'s `:transaction` strategy is basically the equivalent of Rails transactional fixtures, so I guess it makes sense that I would have problems with that too.

I think Machinist was doing some extra transactions or something that interfered with DatabaseCleaner's transactions...

Thanks for the tip about using the latest version from git. Switching from:

```
 gem 'machinist', '>= 2.0.0.beta2'
```

to:

```
 gem 'machinist', :git => 'git://github.com/notahat/machinist.git'
```

fixed the problem for me as well!!
